### DISCUSSION TITLE
Separate color for the 'folder_separator_icon'?

### DISCUSSION THREAD

**Discussion by lujomu** [JanDeDobbeleer/oh-my-posh#2255] [Troubleshoot] [answered]
I'd like the **folder separators** to have a **different color** than the folder names, e.g.
![path_segment_1](https://user-images.githubusercontent.com/8190655/167943565-1166f8d6-e6e2-423e-86a6-b500dab22a4d.png) -> ![path_segment_2](https://user-images.githubusercontent.com/8190655/167943581-b31e6a7d-8a92-4cf2-87df-54b647e012b4.png).

This can be achieved with a **color override** in the `folder_separator_icon` property: `"folder_separator_icon": "<#F9F1A5>\\</>"`.

But in **BASh** (on Windows) this results in **escaping the ASCII escape sequence** and printing `D:\[☻Workbench\[☻Temp`. And if I use `"folder_separator_icon": "<#F9F1A5>\\\\</>"`, it works in BASh, but of course **prints two backslashes** in **PoSh**.

So I have to use **2 distinct path sections** with **conditions** in the **templates**. (See example below.)

Is there a more straight forward way to achieve this, esp. without hardcoding the folder separator symbol?

Example theme:
```
{
  "$schema": "https://raw.githubusercontent.com/JanDeDobbeleer/oh-my-posh/main/themes/schema.json",
  "version": 2,
  "blocks": [
    {
      "type": "prompt",
      "alignment": "left",
      "segments": [
        {
          "type": "path",
          "style": "diamond",
          "leading_diamond": "<parentBackground,background>\uE0B0</>",
          "foreground": "#0C0C0C",
          "background": "#C19C00",
          "template": "{{ if eq .Shell \"pwsh\" }} {{ .Path }} {{ end }}",
          "properties": {
            "style": "agnoster_full",
            "folder_separator_icon": "<#F9F1A5>\\</>"
          }
        },
        {
          "type": "path",
          "style": "diamond",
          "leading_diamond": "<parentBackground,background>\uE0B0</>",
          "foreground": "#0C0C0C",
          "background": "#C19C00",
          "template": "{{ if eq .Shell \"bash\" }} {{ .Path }} {{ end }}",
          "properties": {
            "style": "agnoster_full",
            "folder_separator_icon": "<#F9F1A5>\\\\</>"
          }
        },
        {
          "type": "text",
          "style": "plain",
          "foreground": "parentBackground",
          "background": "transparent",
          "template": "\uE0B0",
          "properties": {}
        }
      ]
    }
  ]
}
```

**Comment by lewis-yeung** [comment, marked as answer]
I had experienced a similar problem before and I think this can be related to:

https://github.com/JanDeDobbeleer/oh-my-posh/blob/6a7f03eb5b215ecae15041a45513f2e68d8a63bf/src/color/ansi.go#L91

In this case for bash, `"folder_separator_icon": "<#F9F1A5>\\</>"` produces something like `...\]\\[...`. As we can see, the folder separator `\` is followed by a `\[` (which is added to mark the beginning of an embedded ANSI color sequence in `$PS1`, refer to [Bash Prompt Escape Sequences](https://tldp.org/HOWTO/Bash-Prompt-HOWTO/bash-prompt-escape-sequences.html) for details) so a `\\[` appears in `$PS1` unexpectedly and it is finally rendered as an interpolated `\[`.

![bug](https://user-images.githubusercontent.com/83903009/167983254-c09970b5-3840-4122-8a6f-b6cd3a57b0fc.png)

When we use double backslashes (`\\`), they are escaped to a single one so everything works correctly.

![normal](https://user-images.githubusercontent.com/83903009/167983658-6c6a5bdd-660b-496f-8f08-9796598ba6c0.png)

TBH, the current logic for escaping special sequences in bash/zsh is not the best solution and indeed it causes problems in some use cases. We may need hard work to refactor it.

**Reply by JanDeDobbeleer** [reply]
@lewis-yeung refactoring that isn't an easy task as I already went through 3 cycles myself. What we have now is the most extensible/comprehensible. It's just a tough nut to crack. I do have a fix for this however.

**Reply by JanDeDobbeleer** [reply]
See #2256

**Reply by lujomu** [reply]
Wow, what a turnaround time! 👏

I can confirm this fixes the issue mentioned above, thanks!

Side note: This does not concern me personally (since I am only using one OS at the moment), but some users might still like to use a **platform specific path separator**. If that need should arise, may I suggest to introduce the properties `folder_separator_foreground` and `folder_separator_background`.

**Reply by JanDeDobbeleer** [reply]
All it would need is template functionality so I'll add that 😃

**Reply by JanDeDobbeleer** [reply]
@lujomu `folder_separator_template` is coming which allows things like this for example: `"<#F9F1A5>{{ if eq .Shell \"bash\"}}/{{ else }}\\{{ end }}</>"` (see #2260)

**Reply by lujomu** [reply]
Works as advertised! Thanks! 🙏

**Reply by QChatagny** [reply]
### Using cycle for background color of subfolder
Hi, not sure its the right place but I'm trying to use cycle to change not only the font color of each subdirectory but the background. Is it possible?  🪴

**Reply by JanDeDobbeleer** [reply]
@QChatagny cycle is an array of color overrides that supports foreground and background colors. See [here](https://ohmyposh.dev/docs/configuration/colors#color-overrides).

**Reply by QChatagny** [reply]
thank you for pointing me out, the documentation is very clear

**Comment by QChatagny** [comment]


**Reply by JanDeDobbeleer** [reply]
Nope.
